src/crypto/hsm/nxp/PersistentStorageOperationalKeystoreHSM.cpp

```diff
@@ -0,0 +1,279 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "PersistentStorageOperationalKeystoreHSM.h"
+#include "CHIPCryptoPALHsm_SE05X_utils.h"
+#include <crypto/OperationalKeystore.h>
+#include <crypto/hsm/CHIPCryptoPALHsm.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <lib/support/SafeInt.h>
+
+#if ENABLE_HSM_GENERATE_EC_KEY
+
+namespace chip {
+
+using namespace chip::Crypto;
+
+namespace {
+
+// Tags for our operational keypair storage.
+constexpr TLV::Tag kOpKeyVersionTag = TLV::ContextTag(0);
+constexpr TLV::Tag kOpKeyDataTag    = TLV::ContextTag(1);
+
+// If this version grows beyond UINT16_MAX, adjust OpKeypairTLVMaxSize
+// accordingly.
+constexpr uint16_t kOpKeyVersion = 1;
+
+constexpr size_t OpKeyTLVMaxSize()
+{
+    // Version and serialized key
+    return TLV::EstimateStructOverhead(sizeof(uint16_t), Crypto::P256SerializedKeypair::Capacity());
+}
+
+} // namespace
+
+#define MAX_KEYID_SLOTS_FOR_FABRICS 32
+#define FABRIC_SE05X_KEYID_START 0x7D100000
+
+struct keyidFabIdMapping_t
+{
+    uint32_t keyId;
+    FabricIndex fabricIndex;
+    bool isPending;
+    Crypto::P256KeypairHSM * pkeyPair;
+} keyidFabIdMapping[MAX_KEYID_SLOTS_FOR_FABRICS] = {
+    0,
+};
+
+uint8_t getEmpytSlotId()
+{
+    uint8_t i = 0;
+    while (i++ < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        if (keyidFabIdMapping[i].keyId == 0 && keyidFabIdMapping[i].isPending == 0)
```

Comment:
And this reads out of bounds on the last iteration.

Reply:
Fixed this.